src/node/blockstorage.cpp

```diff
@@ -679,8 +679,8 @@ bool BlockManager::UndoWriteToDisk(const CBlockUndo& blockundo, FlatFilePos& pos
     }
 
     // Write index header
-    unsigned int nSize = GetSerializeSize(blockundo);
-    fileout << GetParams().MessageStart() << nSize;
+    Assume(blockundo_size == GetSerializeSize(blockundo));
```

Comment:
Calling `Assume(x)` always fully evaluates `x`, so there's no saving here... The only thing `Assume()` does differently in release mode is that it doesn't abort.

Reply:
Damn, you were right, thank you!
Pushed a separate commit that removes the `Assume`s (this way the reviewers and CI can make sure they're the same before/after) - plus found another debug log that has a similar problem, snuck it into this PR as well.

<details>
<summary>Test</summary>

> mkdir demo && cmake -B build -DCMAKE_BUILD_TYPE=Debug && cmake --build build -j$(nproc) && build/src/bitcoind -datadir=demo

```C++
static bool test()
{
    std::cout << "Assume was called!" << std::endl;
    return false;
}

static auto InitBlocksdirXorKey(const BlockManager::Options& opts)
{
    Assume(test());
...
```

prints in both cases, fails only with `Debug`

</details>

Reply:
> Calling `Assume(x)` always fully evaluates `x`, so there's no saving here... The only thing `Assume()` does differently in release mode is that it doesn't abort.

I thought this wasn't the case if `x` can be determined by the compiler to have no side effects?

In any case https://github.com/bitcoin/bitcoin/issues/31178 shows that treating `assume(x)` as a no-op that can be placed in a hot path for free is a bad idea.

Reply:
Migrated the doc update to https://github.com/bitcoin/bitcoin/pull/31528